src/clout/core.cljc

```diff
@@ -0,0 +1,150 @@
+
+(defn- assoc-conj [m k v]
+  (assoc m k
+           (if-let [cur (get m k)]
+             (if (vector? cur)
+               (conj cur v)
+               [cur v])
+             v)))
+
+(defn- assoc-keys-with-groups [groups keys]
+  (reduce (fn [m [k v]] (assoc-conj m k v))
+          {}
+          (map vector keys groups)))
+
+(defn- request-url [request]
+  (str (name (:scheme request))
+       "://"
+       (get-in request [:headers "host"])
+       (:uri request)))
+
+(defn- path-info [request]
+  (or (:path-info request)
+      (:uri request)))
+
+(defprotocol Route
+  (route-matches [route request]
+    "If the route matches the supplied request, the matched keywords are
+    returned as a map. Otherwise, nil is returned."))
+
+(defrecord CompiledRoute [source re keys absolute?]
+  Route
+  (route-matches [_ request]
+    (let [path-info (if absolute?
+                      (request-url request)
+                      (path-info request))]
+      #?(:clj
+         (let [matcher (re-matcher re path-info)]
+           (if (.matches matcher)
+             (assoc-keys-with-groups (re-groups* matcher) keys)))
+         :cljs
+         (if (.test re path-info)
+           (assoc-keys-with-groups (re-groups* re path-info) keys)))))
+  #?@(:clj [Object
+            (toString [_] source)]))
+
+(def ^:private route-parser
+  (insta/parser
+    #?(:clj
+       "route    = (scheme / part) part*
+        scheme   = #'(https?:)?//'
+        <part>   = literal | escaped | wildcard | param
+        literal  = #'(:[^\\p{L}_*{}\\\\]|[^:*{}\\\\])+'
+        escaped  = #'\\\\.'
+        wildcard = '*'
+        param    = key pattern?
+        key      = <':'> #'([\\p{L}_][\\p{L}_0-9-]*)'
+        pattern  = '{' (#'(?:[^{}\\\\]|\\\\.)+' | pattern)* '}'"
+       :cljs
+       "route    = (scheme / part) part*
+        scheme   = #'(https?:)?//'
+        <part>   = literal | escaped | wildcard | param
+        literal  = #'(:[^A-Za-z_*{}\\\\]|[^:*{}\\\\])+'
+        escaped  = #'\\\\.'
+        wildcard = '*'
+        param    = key pattern?
+        key      = <':'> #'([A-Za-z_][A-Za-z_0-9-]*)'
+        pattern  = '{' (#'(?:[^{}\\\\]|\\\\.)+' | pattern)* '}'")
+    :no-slurp true))
+
+(defn- parse [parser text]
+  (let [result (insta/parse parser text)]
+    (if (insta/failure? result)
+      (throw (ex-info "Parse error in route string" {:failure result}))
+      result)))
+
+(defn- find-route-key [form]
+  (case (first form)
+    :wildcard :*
+    :param (-> form second second keyword)))
+
+(defn- route-keys [parse-tree]
+  (->> (rest parse-tree)
+       (filter (comp #{:param :wildcard} first))
+       (map find-route-key)))
+
+(defn- trim-pattern [pattern]
+  (some-> pattern (subs 1 (dec (count pattern)))))
+
+(defn- param-regex [regexs key & [pattern]]
+  (str "(" (or (trim-pattern pattern) (regexs key) "[^/,;?]+") ")"))
+
+(defn- route-regex [parse-tree regexs]
+  (insta/transform
+    {:route    (comp re-pattern #?(:cljs #(str "^" % "$")) str)
```

Comment:
The indentation looks off here. Can you ensure that the code is indented correctly?